titus-ext/aws/src/main/java/com/netflix/titus/ext/aws/iam/AwsIamConnector.java

```diff
@@ -126,6 +153,46 @@ public void shutdown() {
                 });
     }
 
+    @Override
+    public Mono<Void> canAgentAssume(String iamRoleName) {
+        return Mono.defer(() -> {
+            long startTime = registry.clock().wallTime();
+
+            // Check cache first
+            if (Boolean.TRUE.equals(canAssumeCache.getIfPresent(iamRoleName))) {
+                return Mono.empty();
+            }
+
+            // Must call AWS STS service
+            return AwsReactorExt
+                    .<AssumeRoleRequest, AssumeRoleResult>toMono(
+                            () -> new AssumeRoleRequest()
+                                    .withRoleSessionName("titusIamRoleValidation")
+                                    .withRoleArn(iamRoleName)
+                                    .withDurationSeconds(MIN_ASSUMED_ROLE_DURATION_SEC),
+                            stsAgentClient::assumeRoleAsync
+                    )
+                    .flatMap(response -> {
+                        logger.debug("Assumed into: {}", iamRoleName);
+                        canAssumeCache.put(iamRoleName, true);
+                        connectorMetrics.success(IamConnectorMetrics.IamMethods.CanAgentAssume, startTime);
+                        return Mono.<Void>empty();
+                    })
+                    .onErrorMap(error -> {
+                        logger.debug("Error: {}", error.getMessage());
+                        connectorMetrics.failure(IamConnectorMetrics.IamMethods.CanAgentAssume, error, startTime);
+
+                        String errorCode = ((AWSSecurityTokenServiceException) error).getErrorCode();
+                        if ("AccessDenied".equals(errorCode)) {
```

Comment:
You could also consider caching failed validations to prevent frequent re-requests of bad IAMs.

Reply:
Good point. I will add that

Reply:
The downside to this is if the assume permissions are fixed and the value is still cached. If the ttl on the cache is low like 1 minute then it seems fine though.